apps/go-mod-validator/src/parser.ts

```diff
@@ -0,0 +1,76 @@
+const smartContractKitPrefix = "github.com/smartcontractkit/";
+
+function parseGoModLine(line: string) {
+  // Trim any extra whitespace
+  line = line.trim();
+
+  const moduleRegex = new RegExp(
+    /^([\w.-]+(?:\/[\w.-]+)*) (v\d+\.\d+\.\d+(?:-[\w\d.-]+)?(?:\+[\w\d.-]+)?)$/,
+  );
+  const replaceRegex = new RegExp(
+    moduleRegex.source.slice(0, moduleRegex.source.length - 1) +
+      " => " +
+      moduleRegex.source.slice(1),
+  );
+  const moduleMatch = line.match(moduleRegex);
+  const replaceMatch = line.match(replaceRegex);
+
+  if (moduleMatch) {
+    const [, module, version] = moduleMatch;
+    return {
+      module,
+      version,
+    };
+  }
+
+  if (replaceMatch) {
+    const [, , , module, version] = replaceMatch;
+    return {
+      module,
+      version,
+    };
+  }
+
+  return null;
+}
+
+// extract all final dependencies
+function parseGoModLines(lines: string[]) {
```

Comment:
Would it make sense to use [`zx`](https://google.github.io/zx/getting-started) to invoke the `go list -m all` rather than having the output passed in as a file?

Could also do `go list -json -m all` to make it more usable during execution.

Reply:
I initially thought of running `go list -m all` in the script itself then we won't be able to use this script in envs where Go compiler isn't available. Second option I had on mind was to stream the output to the script like this `go list -m all | pnpm run start`, but I wasn't sure how difficult would testing the inputs for such a script be, so finally moved on to passing it as file input. Pls let me know if you think there are better options. 

I also checked `go list -json -m all`, it gives a weird json non-json format, where each module data is listed as json but no the entire output, parsing it was really difficult.

Reply:
> won't be able to use this script in envs where Go compiler isn't available.

This script parses the output of `go list` right? When would the output be available, but the `go` command isn't?
If this is supposed to be run in CI/CD then the user should call `actions/setup-go` first, and if it isn't then you can list `go` as a requirement. 

> I also checked go list -json -m all, it gives a weird json non-json format, where each module data is listed as json but no the entire output, parsing it was really difficult.

Piping the output into `jq` wraps it into an array at least.

```
go list -json -m all | jq -s .
```

Reply:
yes that makes sense. I was thinking of keeping the `go` and `jq` details outside the script. Such that some steps are run before calling our script that actually take care of these details. I feel writing unit tests will also be easy for such scenarios.

Reply:
I've moved `go list` and `jq` commands to the script. One more problems I noticed with redirecting the output of `go list | jq` to the script but then I started hitting shell's cmd line limits. This would be very difficult to guess for each repo. So instead moved the entire thing into the script. 

We still won't be able to build a node action like in `update-action-versions` because we need Go in the running env. Instead I've kept it as a `composite` action.

Tested in https://github.com/smartcontractkit/chainlink-data-streams/actions/runs/9874646427/job/27269498765 and it looks good.

Reply:
Here's a script i whipped up so you dont have to depend on jq. You'll need to convert it to TS but it does work when i tested it on the output of `go list -json -m all > output.txt`

```js
const fs = require("fs");

// We want to parse the following Go data structure into JS objects
// This is the output of `go list -m -json all`
// Example output:
// {
// 	"Path": "github.com/gogo/googleapis",
// 	"Version": "v1.4.1",
// 	"Time": "2021-03-10T14:11:46Z",
// 	"Indirect": true,
// 	"Dir": "/home/parallels/.asdf/installs/golang/1.21.7/packages/pkg/mod/github.com/gogo/googleapis@v1.4.1",
// 	"GoMod": "/home/parallels/.asdf/installs/golang/1.21.7/packages/pkg/mod/cache/download/github.com/gogo/googleapis/@v/v1.4.1.mod",
// 	"GoVersion": "1.12"
// }
//
//  type Module struct {
//         Path       string        // module path
//         Query      string        // version query corresponding to this version
//         Version    string        // module version
//         Versions   []string      // available module versions
//         Replace    *Module       // replaced by this module
//         Time       *time.Time    // time version was created
//         Update     *Module       // available update (with -u)
//         Main       bool          // is this the main module?
//         Indirect   bool          // module is only indirectly needed by main module
//         Dir        string        // directory holding local copy of files, if any
//         GoMod      string        // path to go.mod file describing module, if any
//         GoVersion  string        // go version used in module
//         Retracted  []string      // retraction information, if any (with -retracted or -u)
//         Deprecated string        // deprecation message, if any (with -u)
//         Error      *ModuleError  // error loading module
//         Origin     any           // provenance of module
//         Reuse      bool          // reuse of old module info is safe
//     }

//     type ModuleError struct {
//         Err string // the error itself
//     }

const output = fs.readFileSync("output.txt", "utf8").split("\n");
// store parsed objects
const objects = [];
// keep concatenating lines until the object is complete
let objectString = "";
// keep track of the depth of the object
let objectDepth = 0;
// Process
for (const line of output) {
  if (line.includes("{")) {
    objectDepth++;
  }
  if (line.includes("}")) {
    objectDepth--;
  }

  objectString += line;
  if (objectDepth === 0 && objectString) {
    objects.push(JSON.parse(objectString));
    objectString = "";
  }
}

```

Reply:
Thanks for the script, I've used this one now and removed jq as a dependency